Entry on the BEAM Search algorithm (#2283)
Codecademy/docs: content/ai/concepts/search-algorithms/terms/beam-search/beam-search.md

---
Title: 'BEAM Search'

> Description: 'A graph searching algorithm that is a variant of breadth-first search'

Reviewer: ```suggestion
Description: 'A graph searching algorithm that is a variant of breadth-first search.'
```

Subjects:
  - 'Data Science'
  - 'AI'
Tags:

>   - 'Search Algorithms'

Author: Done!

1. Initialize a search tree with the root node being the start node **S**. This is level zero in the tree.
2. Add **S** to the closed list and evaluate all nodes adjacent to **S**.
3. Select the top β nodes and add them as child nodes to **S** in the tree. Disregard all other nodes.
4. Add the selected nodes to the **open list** for further exploration.
5. Remove all nodes in the **open list**. Once a node is removed from the **open list** for exploration, add it to the **closed list**.
6. Evaluate all adjacent nodes and sort them according to the evaluation function.
7. Select the top β nodes and add them to the tree, connecting them to their respective parent nodes.
8. Repeat this process until the goal node **G** is removed from the open list, indicating that a path has been found.

It is important to note that every level of the search tree contains at most β number of nodes.

## Example

Consider the following example of searching a graph from **S** to **G**. For simplicity, only the edge weights will be used for evaluation. A beam width of β = 2 will be used, meaning only the top two nodes are considered at every level.

> [Example-graph](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-example-graph.png)

Reviewer: ```suggestion
![Example-graph](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-example-graph.png)
```

Initialize the tree and add **S** to the **closed list**:

> [BEAM-Search-1](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-1.png)

Reviewer: ```suggestion
![BEAM-Search-1](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-1.png)
```

**S** only has one adjacent node, so it is added to the tree:

> [BEAM-Search-2](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-2.png)

Reviewer: ```suggestion
![BEAM-Search-2](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-2.png)
```

Explore **I**. Nodes **N** and **H** are the top two nodes, so those are added to the tree:

> [BEAM-Search-3](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-3.png)

Reviewer: ```suggestion
![BEAM-Search-3](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-3.png)
```

Nodes **F** (adjacent to **H** in the graph) and **L** (adjacent to **N** in the graph) are next:

> [BEAM-Search-4](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-4.png)

Reviewer: ```suggestion
![BEAM-Search-4](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-4.png)
```

The incompleteness of this algorithm can be seen in this step. Node **G** has appeared in the **open list** but unfortunately, it has been eliminated by the beam width. Nodes **M** and **C** are next:

> [BEAM-Search-5](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-5.png)

Reviewer: ```suggestion
![BEAM-Search-5](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-5.png)
```

It is in this step that node **G** has been found:

> [BEAM-Search-6](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-6.png)

Reviewer: ```suggestion
![BEAM-Search-6](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-Tree-6.png)
```

Trace the path from **S** to **G**:

> [Final-path](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-final-path.png)

Reviewer: ```suggestion
![Final-path](https://raw.githubusercontent.com/Codecademy/docs/main/media/BEAM-Search-final-path.png)
```
